File: train/connection.py
import redis
import time
import numpy as np
import json


class Connection:
    def __init__(self):
        self.r = redis.StrictRedis(host='localhost', port=6379)
        self.p = self.r.pubsub()
        self.p.subscribe('state_buffer')

    def get_state_buffer(self):
        while True:
            msg = self.p.get_message()
            if msg:
                if type(msg['data']) == bytes:
                    s = msg['data'].decode('utf-8')
                    d = json.loads(s)
                    a = np.array(d['data'], dtype='float32')
                    return a
            time.sleep(0.001)

    def publish_action(self, action):
        self.r.publish('action', list(action))


# Test
if __name__ == '__main__':
    connection = Connection()
    while True:
        connection.publish_action(np.array([1.0, 0.0, 0.0, 0.0, 0.0]))
        print(connection.get_state_buffer())
        time.sleep(0.0001)

File: train/connection.js
const redis = require('redis');

class Connection {
  constructor() {
    this.pub = redis.createClient();
    this.sub = redis.createClient();
  }

  set_action_callback(action_callback) {
    this.sub.on('message', function(channel, message) {
      try {
        const action = JSON.parse(message);
        action_callback(action);
      } catch (e) {
        console.log("ERROR: Incorrectly formatted message.");
      }
    })
    this.sub.subscribe('action');
  }

  publish_state_buffer(state_buffer) {
    this.pub.publish('state_buffer', JSON.stringify({
      data: state_buffer
    }))
  }
}

module.exports = {
  Connection: Connection
}
